README.md
<h1 align="center">
  <img src="https://fakeimg.pl/900x300/ffffff/333333/?text=coerce+middleware&font=museo" alt="coerce middleware" width="900px" />
</h1>

<p align="center">coerce req.query, req.params or req.body</p>

<p align="center">
<a href="https://opensource.org/licenses">
  <img src="https://img.shields.io/github/license/ezylean/coerce-middleware.svg" alt="License" />
</a>
<a href="https://circleci.com/gh/ezylean/coerce-middleware/tree/master">
  <img src="https://circleci.com/gh/ezylean/coerce-middleware/tree/master.svg?style=shield" alt="CircleCI" />
</a>
<a href="https://codecov.io/gh/ezylean/coerce-middleware">
  <img src="https://codecov.io/gh/ezylean/coerce-middleware/branch/master/graph/badge.svg" alt="codecov" />
</a>
<a href="https://ezylean.github.io/coerce-middleware">
  <img src="https://img.shields.io/badge/docs-typedoc-%239B55FC.svg" alt="Docs: typedoc" />
</a>
<a href="https://github.com/ezylean/coerce-middleware/issues">
  <img src="https://img.shields.io/github/issues-raw/ezylean/coerce-middleware.svg" alt="GitHub issues" />
</a>
<a href="https://codeclimate.com/github/ezylean/coerce-middleware/maintainability" >
  <img src="https://img.shields.io/codeclimate/maintainability-percentage/ezylean/coerce-middleware.svg" alt="Maintainability" />
</a>
<a href="https://david-dm.org/ezylean/coerce-middleware">
  <img src="https://david-dm.org/ezylean/coerce-middleware.svg" alt="Dependencies status" />
</a>
<a href="https://david-dm.org/ezylean/coerce-middleware?type=dev">
  <img src="https://david-dm.org/ezylean/coerce-middleware/dev-status.svg" alt="Dev Dependencies status" />
</a>
<a href="https://github.com/Microsoft/TypeScript">
  <img src="https://img.shields.io/badge/made%20with-typescript-%234B9DD5.svg" alt="Made with: typescript" />
</a>
<a href="https://github.com/prettier/prettier">
  <img src="https://img.shields.io/badge/code%20style-prettier-ff69b4.svg" alt="Code style: prettier" />
</a>
<a href="https://npm.runkit.com/@ezy/coerce-middleware">
  <img src="https://img.shields.io/badge/runkit-try%20now-%236967CA.svg" alt="Runkit: try now" />
</a>
<a href="https://bundlephobia.com/result?p=@ezy/coerce-middleware">
  <img src="https://img.shields.io/bundlephobia/min/@ezy/coerce-middleware.svg" alt="minified size" />
</a>
<a href="https://bundlephobia.com/result?p=@ezy/coerce-middleware">
  <img src="https://img.shields.io/bundlephobia/minzip/@ezy/coerce-middleware.svg" alt="minzipped size" />
</a>
</p>

## Why

Validation logic become simpler when it's possible to expect "typed" datas, but datas parsed from url like req.query or req.params are strings.

## Install

```shell
npm i @ezy/coerce-middleware
```

## Usage

### Common

```js
import { coerce } from '@ezy/coerce-middleware';
import express from 'express';
const app = express();

app.use(coerce('query'));

app.get('/', (req, res) => {
  res.send(JSON.stringify(req.query));
});

app.listen(3000);

// GET /?id=30&is_admin=true&name=awesome%20title
// => {
//   name: 'awesome title',
//   id: 30,
//   is_admin: true
// }
```

### use a custom coercePrimitive function

```js
import { coerce } from '@ezy/coerce-middleware';
import express from 'express';
const app = express();

app.use(coerce('query', value => '(^^)'));

app.get('/', (req, res) => {
  res.send(JSON.stringify(req.query));
});

app.listen(3000);

// GET /?id=30&is_admin=true&nested[name]=awesome%20title
// => {
//   nested: {
//      name: '(^^)'
//   },
//   id: '(^^)',
//   is_admin: '(^^)'
// }
```

## Links

- [API docs](https://ezylean.github.io/coerce-middleware)
- [Playground](https://npm.runkit.com/@ezy/coerce-middleware)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.1.1"></a>
## [2.1.1](https://github.com/ezylean/coerce-middleware/compare/v2.1.0...v2.1.1) (2019-09-22)


### Reverts

* **package.json:** revert package private ([b41842e](https://github.com/ezylean/coerce-middleware/commit/b41842e))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/ezylean/coerce-middleware/compare/v2.0.1...v2.1.0) (2019-09-22)


### Features

* **defaultCoercePrimitive:** add support for Date conversion ([8aa5a0a](https://github.com/ezylean/coerce-middleware/commit/8aa5a0a))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/ezylean/coerce-middleware/compare/v2.0.0...v2.0.1) (2018-09-25)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/ezylean/coerce-middleware/compare/v1.0.1...v2.0.0) (2018-09-24)


### Performance Improvements

* **dependencies:** replace overkill 'deep-map' with an in-house solution ([36b6b2c](https://github.com/ezylean/coerce-middleware/commit/36b6b2c))


### BREAKING CHANGES

* **dependencies:** properties setted to undefined are not preserved anymore



<a name="1.0.1"></a>
## [1.0.1](https://github.com/ezylean/coerce-middleware/compare/v1.0.0...v1.0.1) (2018-09-23)



<a name="1.0.0"></a>
# 1.0.0 (2018-09-23)


### Features

* **init:** first version ([4a7db0c](https://github.com/ezylean/coerce-middleware/commit/4a7db0c))


src/lib/coerce.spec.ts
// tslint:disable:no-expression-statement
import test from 'ava';
import { IncomingMessage, ServerResponse } from 'http';

import { coerce } from './coerce';

test('coerce stringified values', t => {
  const mw = coerce('query');

  const req: IncomingMessage & { query: any } = {
    query: {
      aborted: 'false',
      agree_terms: 'true',
      longDate: '2019-09-22T13:23:05.254Z',
      name: 'lorem ipsum',
      owner_id: '56',
      price: '24.99',
      shortDate: '2019-09-22'
    }
  } as any;

  const expected = {
    aborted: false,
    agree_terms: true,
    longDate: new Date('2019-09-22T13:23:05.254Z'),
    name: 'lorem ipsum',
    owner_id: 56,
    price: 24.99,
    shortDate: new Date('2019-09-22')
  };

  const res: ServerResponse = {} as any;
  const next = (e?: Error) => {
    t.deepEqual(e, undefined);
    t.deepEqual(req.query, expected);
  };

  mw(req, res, next);
});

test('coerce messy object', t => {
  const mw = coerce('body');

  const req: IncomingMessage & { body: any } = {
    body: {
      distance: '125.86',
      id: 12,
      name: 'lorem ipsum',
      someOtherField: 'null'
    }
  } as any;

  const expected = {
    distance: 125.86,
    id: 12,
    name: 'lorem ipsum',
    someOtherField: null
  };

  const res: ServerResponse = {} as any;
  const next = (e?: Error) => {
    t.deepEqual(e, undefined);
    t.deepEqual(req.body, expected);
  };

  mw(req, res, next);
});

test('use custom coercePrimitive function', t => {
  const coercePrimitive = () => {
    throw new Error('oops');
  };

  const mw = coerce('body', coercePrimitive);

  const body = {
    id: 10,
    name: 'name'
  };

  const req: IncomingMessage & { body: any } = {
    body
  } as any;

  const expected = new Error('oops');

  const res: ServerResponse = {} as any;
  const next = (e?: Error) => {
    t.deepEqual(e, expected);
    t.deepEqual(req.body, body);
  };

  mw(req, res, next);
});

test('dont change anything on primitive/undefined property', t => {
  const mw = coerce('method');

  const req: IncomingMessage = { method: 'GET' } as any;
  const res: ServerResponse = {} as any;

  const next = (e?: Error) => {
    t.deepEqual(e, undefined);
    t.deepEqual(req, { method: 'GET' } as any);
  };

  mw(req, res, next);

  const req2: IncomingMessage = {} as any;
  const res2: ServerResponse = {} as any;

  const next2 = (e?: Error) => {
    t.deepEqual(e, undefined);
    t.deepEqual(req2, {} as any);
  };

  mw(req2, res2, next2);
});


assets/js/search.js
var typedoc = typedoc || {};
            typedoc.search = typedoc.search || {};
            typedoc.search.data = {"kinds":{"64":"Function"},"rows":[{"id":0,"kind":64,"name":"coerce","url":"globals.html#coerce","classes":"tsd-kind-function"},{"id":1,"kind":64,"name":"isPrimitive","url":"globals.html#isprimitive","classes":"tsd-kind-function tsd-is-not-exported"},{"id":2,"kind":64,"name":"deepMap","url":"globals.html#deepmap","classes":"tsd-kind-function tsd-is-not-exported"},{"id":3,"kind":64,"name":"isInt","url":"globals.html#isint","classes":"tsd-kind-function tsd-is-not-exported"},{"id":4,"kind":64,"name":"isDecimal","url":"globals.html#isdecimal","classes":"tsd-kind-function tsd-is-not-exported"},{"id":5,"kind":64,"name":"defaultCoercePrimitive","url":"globals.html#defaultcoerceprimitive","classes":"tsd-kind-function tsd-is-not-exported"}]};

src/lib/coerce.ts
// tslint:disable:prefer-for-of
import { from as fromDesc, to as toDesc } from '@ezy/object-description';
import { IncomingMessage, ServerResponse } from 'http';

/**
 * Generate a connect-style middleware that coerce values for a req property
 *
 * ### Example
 * ```js
 * import { coerce } from '@ezy/coerce-middleware'
 * import express from 'express'
 * const app = express()
 *
 * app.use(coerce('query'))
 *
 * app.get('/', (req, res) => {
 *   res.send(JSON.stringify(req.query))
 * })
 *
 * app.listen(3000)
 *
 * // GET /?id=30&is_admin=true&name=awesome%20title
 * // => {
 * //   name: 'awesome title',
 * //   id: 30,
 * //   is_admin: true
 * // }
 * ```
 *
 * @param property            req property to coerce (eg: 'query', 'params', etc...)
 * @param coercePrimitive     an alternative function to coerce stringified primitives
 * @returns                   a connect-style middleware
 */
export function coerce(
  properties: string | string[],
  coercePrimitive: (
    value: string
  ) => string | number | boolean | Date | void = defaultCoercePrimitive
) {
  return (
    req: IncomingMessage,
    // @ts-ignore
    res: ServerResponse,
    next: (e?: Error) => void
  ) => {
    if (!Array.isArray(properties)) {
      properties = [properties];
    }

    for (let index = 0; index < properties.length; index++) {
      const property = properties[index];
      if (isPrimitive(req[property])) {
        return next();
      }

      try {
        req[property] = deepMap(
          req[property],
          value => (typeof value === 'string' ? coercePrimitive(value) : value)
        );

        next();
      } catch (error) {
        next(error);
      }
    }
  };
}

/**
 * check if a given value is a javascript primitive
 */
function isPrimitive(value) {
  return value !== Object(value);
}

/**
 * a map-like function for primitives in nested object or array
 *
 * @param object    any object/array.
 * @param mapFn     function to apply on primitive values.
 * @returns         the mapped object/array.
 */
function deepMap(object, mapFn: (value) => any): any {
  const desc = toDesc(object);

  const values = desc.values.map(primitive => {
    primitive.value = mapFn(primitive.value);
    return primitive;
  });

  return fromDesc({ ...desc, values });
}

/**
 * coerce a stringified primitive value
 *
 * @param value   any stringified primitive value.
 * @returns       the same value type converted.
 */
function defaultCoercePrimitive(value: string) {
  if (value) {
    if (value === 'null') {
      return null;
    }

    if (value === 'true') {
      return true;
    }

    if (value === 'false') {
      return false;
    }

    const num = Number(value);
    if (!isNaN(num)) {
      return num;
    }

    const date = new Date(value);
    if (!isNaN(date as any)) {
      return date;
    }
  }

  return value;
}
